libsolidity/analysis/FunctionCallGraph.cpp

```diff
@@ -0,0 +1,324 @@
+{
+	if (_lhs.index() != _rhs.index())
+		return _lhs.index() < _rhs.index();
+
+	if (std::holds_alternative<SpecialNode>(_lhs))
+		return std::get<SpecialNode>(_lhs) < std::get<SpecialNode>(_rhs);
+	return std::get<ASTNode const*>(_lhs)->id() < std::get<ASTNode const*>(_rhs)->id();
+}
+
+bool FunctionCallGraphBuilder::CompareByID::operator()(Node const& _lhs, int64_t _rhs) const
+{
+	solAssert(!std::holds_alternative<SpecialNode>(_lhs), "");
+
+	return std::get<ASTNode const*>(_lhs)->id() < _rhs;
+}
+
+bool FunctionCallGraphBuilder::CompareByID::operator()(int64_t _lhs, Node const& _rhs) const
+{
+	solAssert(!std::holds_alternative<SpecialNode>(_rhs), "");
+
+	return _lhs < std::get<ASTNode const*>(_rhs)->id();
+}
+
+unique_ptr<FunctionCallGraphBuilder::ContractCallGraph> FunctionCallGraphBuilder::create(ContractDefinition const& _contract)
+{
+	return FunctionCallGraphBuilder(_contract).m_graph;
+}
+
+bool FunctionCallGraphBuilder::visit(FunctionCall const& _functionCall)
+{
+	solAssert(m_currentNode.has_value(), "");
+	solAssert(holds_alternative<SpecialNode>(*m_currentNode) || get<ASTNode const*>(*m_currentNode) != nullptr, "");
+
+	FunctionType const* functionType = dynamic_cast<FunctionType const*>(_functionCall.expression().annotation().type);
+	if (
+		functionType &&
+		functionType->kind() == FunctionType::Kind::Internal &&
+		!_functionCall.expression().annotation().calledDirectly
+	)
+		// If it's not a direct call, we don't really know which function will be called (it may even
+		// change at runtime). All we can do is to add an edge to the dispatch which in turn has
+		// edges to all functions could possibly be called.
+		add(*m_currentNode, m_currentDispatch);
+	else if (functionType && functionType->kind() == FunctionType::Kind::Event)
+	{
+		// NOTE: calledDirectly is always false for events, no matter whether it's an actual emit
+		// or just event name used in an expression. Does not matter since this visit(FunctionCall)
+		// won't get triggered in the latter case.
+		solAssert(!_functionCall.expression().annotation().calledDirectly, "");
+
+		EventDefinition const* event = nullptr;
+		if (auto memberAccess = dynamic_cast<MemberAccess const*>(&_functionCall.expression()))
+		{
+			solAssert(*memberAccess->annotation().requiredLookup == VirtualLookup::Static, "");
+			event = dynamic_cast<EventDefinition const*>(memberAccess->annotation().referencedDeclaration);
+		}
+		else if (auto identifier = dynamic_cast<Identifier const*>(&_functionCall.expression()))
+		{
+			// TMP: Why is the lookup type virtual for when we refer to the event just by name
+			// and static when we qualify it with the contract name?
+			solAssert(*identifier->annotation().requiredLookup == VirtualLookup::Virtual, "");
+			event = dynamic_cast<EventDefinition const*>(identifier->annotation().referencedDeclaration);
+		}
+		else
+			solAssert(false, "");
+
+		solAssert(event, "");
+		add(*m_currentNode, event);
```

Comment:
Please collect the set of events similar to `createdContracts`. I don't think we need an edge to the event.

Reply:
ok